### PR TITLE
Improved DoF widget

### PR DESCRIPTION
**Pull request by dvr333**
Incorporated a separable Frostbite 2-style DoF implementation into the DoF widget.

### CODE REVIEW

File: LuaUI/Widgets/Shaders/dof.fs [hunk 1]
```diff
@@ -90,7 +90,7 @@ float LinearizeDepth(vec2 uv){
 
     float n22 = viewProjectionInv[2][2];
```
Comment (lhog) [inline LuaUI/Widgets/Shaders/dof.fs]: The formula below expects projection matrix, not the inverse of (viewProj)

Comment (dvr333) [reply LuaUI/Widgets/Shaders/dof.fs]: Oh, right, the name is misleading. If you look at the Lua you'll see it's taking the "projection" matrix now, but I didn't want to deal with renames at 2AM.

Comment (lhog) [reply LuaUI/Widgets/Shaders/dof.fs]: Sorry missed that.

**Comment by sprunk** [comment]
Got a comparison screenshot?

File: LuaUI/Widgets/Shaders/dof.fs (hunk 1 above)
Comment (sprunk) [reply LuaUI/Widgets/Shaders/dof.fs]: At least comment in the code would be good.

Comment (lhog) [reply LuaUI/Widgets/Shaders/dof.fs]: I think the @dvr333 answer implies he's going to fix name when it's not 2 AM :)

Comment (dvr333) [reply LuaUI/Widgets/Shaders/dof.fs]: It's no longer 2AM, and the problem is fixed, along with a more pressing problem of the correctness of the blur, both at the edges of in-focus objects and in general. Turns out glsl's "inout" is the equivalent of C#'s "out", which threw me off.

Comment (lhog) [reply LuaUI/Widgets/Shaders/dof.fs]: Not quite equivalent. All values are still copied (not passed by reference), so `inout` means an input and output copy of value is made. `out` means only output copy is made. `in` AFAIK has no separate meaning from just default behavior (input copy is made)

**Comment by dvr333** [comment]
> 
> 
> Got a comparison screenshot?

Here's an [album](https://imgur.com/a/0y10VbT). The first 8 shots are more artistic shots to really show off the effect (I did actually have the old effect on in all the shots that look normal, but it's really subtle), the last 4 are taken from angles that are representative of how the camera will be placed in-game while playing normally.

**Comment by GoogleFrog** [comment]
I'm adding WIP until it is clear that this is ready. Make sure you can rename it, and rename it when it is ready.

**Comment by dvr333** [comment]
I can't rename it, but I'll put a comment here when I consider it ready. It's pretty close now. I have an experiment to improve the look of objects nearer to the camera than the focus plane, and depending on that I'll decide whether autofocus should be realistic (success) or more of a tilt-shift effect (failure).

**Comment by sprunk** [comment]
@dvr333 pinging to ensure you're aware of the comment above.

**Comment by dvr333** [comment]
Yep, I got it. I'm going to do a bit of testing and tweaking over the weekend (though to be fair working on this did hurt my hands a bit more than expected), if it's done, the "WIP" will come off.

**Comment by dvr333** [comment]
Sorry, misclicked on the "Close and Comment" button.

**Comment by dvr333** [comment]
Todo before removing WIP:

- [ ] Ensure effect is resolution-independent
- [ ] Add proper blur for near-to-camera objects

**Comment by dvr333** [comment]
Currently debating whether to actually have the proper blur for near-to-camera objects. Currently objects nearer to the camera than the focus plane don't look half bad. It's not perfect, but it's still pretty clearly a blur. I might give myself a couple days to see if I can work out why my current attempts aren't working, but if I can't I'll just take off the WIP anyway, as this near-to-camera object blur is intended as a "high quality" default-off option anyhow, and I don't see the point of holding up releasing this on an optional default-off feature.

**Pull request merged**